langchain/agents/agent_toolkits/openapi/planner.py

```diff
@@ -0,0 +1,375 @@
+Some user queries can be resolved in a single API call, but some will require several API calls.
+You will generate a plan of API calls and explain what they are accomplishing step by step.
+The plan will be passed to an API controller that can format them into web requests and return the responses.
+
+----
+
+Endpoints:
+
+{endpoints}
+
+----
+
+Suppose we're interested in Spotify's API:
+
+Good examples:
+
+User query: Find me songs by Norah Jones.
+Plan:
+1. GET /search to get the artist ID for Norah Jones.
+2. GET /artists/{{id}}/top-tracks to get the artist's top-tracks.
+
+User query: Give me some jazz recommendations.
+Plan:
+1. GET /recommendations/available-genre-seeds to find valid recommendation seeds.
+2. GET /recommendations to find recommendations.
+
+Bad examples:
+
+User query: Add the song Central Park West to my queue
+Plan:
+1. POST /me/player/queue to add the track to the user's queue
+Explanation: This plan did not look up the track id for the song. It also did not look up information about the user.
+
+----
+
+User query: {query}
+Plan:"""
+API_PLANNER_TOOL_NAME = "api_planner"
+API_PLANNER_TOOL_DESCRIPTION = f"Can be used generate the right API calls to assist with a user query, like {API_PLANNER_TOOL_NAME}(query). Should always be called before trying to calling the API controller."
+
+# Execution.
+API_CONTROLLER_PROMPT = """You are an agent that gets a sequence of API calls and given their documentation, should execute them and return the final response.
+If you cannot complete them and run into issues, you should explain the issue. When interacting with API objects, you should extract ids for inputs to other API calls but ids and names for outputs returned to the User.
+
+
+Here is documentation on the API:
+Base url: {api_url}
+Endpoints:
+{api_docs}
+
+
+Here are tools to execute requests against the API: {tool_descriptions}
+
+
+Starting below, you should follow this format:
+
+Plan: the plan of API calls to execute
+Thought: you should always think about what to do
+Action: the action to take, should be one of the tools [{tool_names}]
+Action Input: the input to the action
+Observation: the output of the action
+... (this Thought/Action/Action Input/Observation can repeat N times)
+Thought: I am finished executing the plan (or, I cannot finish executing the plan without knowing some other information.) 
+Final Answer: the final output from executing the plan or missing information I'd need to re-plan correctly.
+
+
+You can use US as a country code.
+
+Begin!
+
+Plan: {input}
+Thought:
+{agent_scratchpad}
+"""
+API_CONTROLLER_TOOL_NAME = "api_controller"
+API_CONTROLLER_TOOL_DESCRIPTION = (
+    f"Can be used execute a plan of API calls, like {API_CONTROLLER_TOOL_NAME}(plan)."
+)
+
+# Orchestrate planning + execution.
+API_ORCHESTRATOR_PROMPT = """You are an agent that assists with user queries against API, things like querying information or creating resources.
```

Comment:
Could you leave a comment somewhere to explain why building an orchestration agent is needed? (vs. composing two llm chains?)

Reply:
Added and also had this documented a bit below. 

Composing/chaining is a reasonable idea... 

I started with planner as being responsible for generating a plan as well as invoking the execution, i.e. it was an agent. There were too many instructions stuffed into the prompt so I split it out. However having an agent at the top-level is nice... it can generate a plan, try to execute it, have errors bubble up and re-plan. Not that I do this in any repeatable, sophisticated way atm, but it's a natural next step.